### ISSUE TITLE
Can not read file name with umlauts in boost::beast::http::file_body

### ISSUE THREAD

**Issue by emmenlau** [boostorg/beast#2081]
Thanks again for the very cool library, first and foremost! We have a minor issue with umlauts (low priority).

### Version of Beast

```
#define BOOST_BEAST_VERSION 290
```

### Steps necessary to reproduce the problem

My code is based on the advanced_server_flex.cpp example, and my problem appears specifically around line
https://github.com/boostorg/beast/blob/develop/example/advanced/server-flex/advanced_server_flex.cpp#L179:
```cpp
    // Attempt to open the file
    beast::error_code ec;
    http::file_body::value_type body;
    body.open(path.c_str(), beast::file_mode::scan, ec);

    // Handle the case where the file doesn't exist
    if(ec == beast::errc::no_such_file_or_directory)  //< This is true for us when the file name has umlauts.
```

We used this code successfully on multiple platforms (MSVC, Ubuntu and MacOSX) for about a year now. However as soon as we started adding `clang-cl` to the list of compilers, the webserver can no longer read files that have Umlauts in the file name.

We tested with a file named
```
üöälaut.txt
```
that is _not_ found on Windows when using clang-cl, whereas all other platforms/compilers work fine.

### All relevant compiler information

The error comes with Visual Studio 2019 x64 16.7.1 with the Microsoft-supplied `clang-cl` version 10.0.0. Using the Microsoft-Compiler and not clang-cl, everything works.

**Comment by madmongo1** [comment]
Thanks for reporting this @emmenlau ,

Before I start looking into this, do you have the impression that this is a URL parsing problem or a file access problem?

**Comment by emmenlau** [comment]
> Thanks for reporting this @emmenlau ,
> 
> Before I start looking into this, do you have the impression that this is a URL parsing problem or a file access problem?

Thanks a lot @madmongo1 ! Admittedly I don't know.

The file was created with boost::filesystem on all platforms, and can be correctly queried with boost::filesystem on all platforms. Also, Windows Explorer shows the file correctly. So the least I can guess, its not a problem with the file not existing or being inaccessible due to permissions.

The server also prints the file name correctly when I print the std::string filename:
```
handle_request(): The resource 'C:/data/stable-tmp-MSVC-Haswell-7-x64-cl19.27.29111_clang10.0.0/Debug/public/üöälaut.txt' was not found.
```

Does that help? Let me know anything I can do to further narrow this down...

**Comment by vinniefalco** [comment]
probably having to do with `fopen` and utf-8 encoded paths or something like that

**Comment by emmenlau** [comment]
> probably having to do with `fopen` and utf-8 encoded paths or something like that

Yes that sounds about right. We encode paths as UTF-8, and now that you mention it, I remember that boost::filesystem required a default conversion for Windows to be set. Something along the lines of
```
boost::filesystem::path::imbue( 
    std::locale( std::locale(), new std::codecvt_utf8_utf16<wchar_t>() ) );
```

Its interesting however that this used to work with boost::beast and the MSVC compilers out of the box, whereas clang-cl seems to treat it differently.

**Comment by vinniefalco** [comment]
ugh, facets....

**Comment by vinniefalco** [comment]
`beast::file` is _supposed_ to handle utf-8 encoded filenames, but I never implemented it.

**Comment by emmenlau** [comment]
> `beast::file` is _supposed_ to handle utf-8 encoded filenames, but I never implemented it.

Hehe, fair enough! Do you think this will be on the road map for the next 3-6 months? Or is it rather unlikely to happen? The issue has no urgency for us, but it would be nice to know that this will come at some point...

**Comment by vinniefalco** [comment]
> Do you think this will be on the road map for the next 3-6 months?

@madmongo1 ?

**Comment by madmongo1** [comment]
We have a user with an unimplemented use case. My pride will not allow me to leave this wrong unrighted.
It’s going to the top of the development queue.

**Comment by emmenlau** [comment]
> It’s going to the top of the development queue.

Hahaha noooo! Its really not so urgent! :-) But guys, thanks again for the quick response!

**Comment by madmongo1** [comment]
Hi @emmenlau, would you mind checking whether adding a manifest to your executable solves this?

It will still need a code solution for older versions of windows, but I am interested to know whether this works:

https://docs.microsoft.com/en-us/windows/uwp/design/globalizing/use-utf8-code-page

**Comment by emmenlau** [comment]
> Hi @emmenlau, would you mind checking whether adding a manifest to your executable solves this?
> 
> It will still need a code solution for older versions of windows, but I am interested to know whether this works:
> 
> https://docs.microsoft.com/en-us/windows/uwp/design/globalizing/use-utf8-code-page

This is awesome, I was not aware of it! Sadly, we run all our builds and tests on Windows 7 so I can't test it (yet). But this is a very interesting new insight for me. Nice work from Microsoft.

**Comment by madmongo1** [comment]
Hi @emmenlau ,

Are you able to give me the sequence of bytes you think you are getting for `"üöälaut.txt"` ?

I have performed a conversion online to : `{ 0xc3, 0xbc, 0xc3, 0xb6, 0xc3, 0xa4, 0x6c, 0x61, 0x75, 0x74, 0x2e , 0x74, 0x78, 0x74 };`

But this is not converting correctly to UTF-16 via `::MultiByteToWideChar`

**Comment by emmenlau** [comment]
Argh! I should be more helpful in this, but sadly I don't know how to get this. Originally the sequence is encoded in UTF-8, but this is not what you want, no? Then we use this UTF-8 string in boost::filesystem to create test files that boost::beast should serve. The client sends the UTF-8 string via a boost::beast-based HTTP connection to the server. I assume what we receive server-side is still UTF-8, and when I print it, it looks correct. So far, also everything works fine with any compiler I tried.

The error seems to come when I pass this (probably UTF-8 encoded) string to `body.open(path.c_str(), beast::file_mode::scan, ec);`: While the file reading works with all other compilers, it fails with `clang-cl`, claiming the file would not exist. But I have no way of tracking how clang translates/encodes/parses the string, and also no way of knowing what exactly fails in opening the file.

I can only say that in boost::filesystem, everything works when we `imbue` the automatic conversion with `std::codecvt_utf8_utf16<wchar_t>()`.

Cutting a long story short, what can I do to be more helpful? Should I try to step into this with the debugger to see if I can find more details? Or does `body.open()` accept UTF-16 so I can try to perform a conversion manually?

**Comment by madmongo1** [comment]
Thanks for responding @emmenlau 

What would really help me track this down is:

1. a simple 10-line program that demonstrates this causing a problem on clang-cl but not on msvc or gcc.
2. idiot-proof instructions on how to compile and link it :-)

This way I can replicate without too much effort and focus my attention on what is going wrong. It's beginning to sound as if we're not detecting Windows properly when compiling on clang, but I'd like to be able to prove it.

R

**Comment by emmenlau** [comment]
Dear @madmongo1 ,

sorry for the super late reply, I did not have time to continue with the issue. Now I've just created a relatively small test-case that inhibits the aforementioned problem. Its based on gtest, boost::filesystem to create a test file, and boost::beast to open the test file. The test works for me in our CI-system on Linux, macOS and the Microsoft Compiler. It only fails (as expected) when using ClangCl on MSVC. Feel free to throw out gtest if you need, its not necessarily required.

Can you try to reproduce the problem on your side? Or do you spot any obvious mistakes in the test?

Here the file for download:
[BoostBeastClangCLUmlautsTest.txt](https://github.com/boostorg/beast/files/5328189/BoostBeastClangCLUmlautsTest.txt)

**Comment by madmongo1** [comment]
Fantastic. Thank you.

**Comment by emmenlau** [comment]
Hi all, just a friendly ping on this issue with Ümlauts... :-)

**Comment by madmongo1** [comment]
Thanks. Will take a look.

**Comment by emmenlau** [comment]
Finally I can maybe shed some light on this issue. When using the clang-cl compiler, it is currently detected by `boost/config.hpp` as clang, not MSVC. Therefore the define `BOOST_MSVC` is missing. This brings along the problem that boost::beast does not use the `wchar` APIs for the Windows file interface.

**Comment by vinniefalco** [comment]
@pdimov what do we do

**Comment by emmenlau** [comment]
Progressing a bit further with the investigation, maybe boost::config is correct to include the header `clang.hpp`. I come to this idea because `clang.hpp` has some support for clang-cl. It uses `_MSC_VER` in a few places. If boost::config is correct here, then maybe `BOOST_MSVC` is not the only way to detect compilation based on Visual Studio? Another possibility seems to be something like `#if defined(BOOST_CLANG) && defined(_MSC_VER)`.

**Comment by vinniefalco** [comment]
or just `#ifdef _MSC_VER`

**Comment by pdimov** [comment]
> @pdimov what do we do

fix the code to not be broken

**Comment by pdimov** [comment]
https://github.com/boostorg/beast/blob/b15a5ff0e47e72ba3d4711d2514bc65749d036ae/include/boost/beast/core/file_win32.hpp#L16-L20

This uses the Win32 file implementation only on MSVC, but there exist other compilers targeting the Windows platform. It should probably detect the platform and not the specific compiler, by using `_WIN32`, for instance.

**Comment by pdimov** [comment]
`_MSC_VER` will also work for `clang-cl` and e.g. Borland, but it will not pick up Cygwin GCC and MinGW GCC.

**Comment by emmenlau** [comment]
> https://github.com/boostorg/beast/blob/b15a5ff0e47e72ba3d4711d2514bc65749d036ae/include/boost/beast/core/file_win32.hpp#L16-L20
> 
> This uses the Win32 file implementation only on MSVC, but there exist other compilers targeting the Windows platform. It should probably detect the platform and not the specific compiler, by using `_WIN32`, for instance.

Oh I did not even see `BOOST_BEAST_USE_WIN32_FILE` before. In the places I stepped through file operations are bound to `BOOST_MSVC` and not to `BOOST_BEAST_USE_WIN32_FILE`. So maybe the following should use `BOOST_BEAST_USE_WIN32_FILE` where currently `BOOST_MSVC` is used?

https://github.com/boostorg/beast/blob/b7344b0d501f23f763a76488826dde8c31f34b5d/include/boost/beast/core/impl/file_stdio.ipp#L84-L91

**Comment by pdimov** [comment]
Yes, all these should also check for whether the Microsoft C library is used, and not the specific compiler. `_MSC_VER` is probably the right thing here.

I'm not sure when the stdio implementation will be used under Windows though, instead of the Win32 one. When BOOST_BEAST_USE_WIN32_FILE is defined by the user to 0, perhaps.

**Comment by emmenlau** [comment]
I just wanted to kindly ping here to see if this is still on the horizon? I'm a bit scared now when using beast with clang, due to the possibly wrong defines that may break the code in arbitrary ways... but clang is such a great compiler front-end on MSVC nowadays...

**Comment by madmongo1** [comment]
No progress has been made I'm afraid. Do you have a set of defines that causes the correct behaviour with the clang front end?

**Comment by pdimov** [comment]
https://github.com/boostorg/beast/pull/2353 should fix the problem in the `file_stdio` implementation.

**Comment by pdimov** [comment]
https://github.com/boostorg/beast/pull/2354 will enable `file_win32` on Windows, instead of just on MSVC.

**Comment by emmenlau** [comment]
> Do you have a set of defines that causes the correct behaviour with the clang front end?

No, sadly not. I'm actually quite scared to mess with the boost defines for clang-cl. I know that in many aspects, clang-cl emulates MSVC behavior. But in other aspects it behaves like clang. And possibly it can do both at the same time.

Do you know if there is a "compiler support team" in boost that can look into cleanly integrating clang-cl first? I understand that currently clang-cl is seeing the clang header instead of the MSVC header, and this could be a problem in itself...

**Comment by sehe** [comment]
Late bit of feedback out of the field from a [related SO question](https://stackoverflow.com/questions/71907767/how-to-create-a-response-from-a-file-that-has-cjk-filename/72093070?noredirect=1#comment127531785_72093070): 

> Sorry, I already replaced the file body with a string body in my codes. To change the design, I may have to modify some other codes. Maybe I will try it in other projects in the future. Anyway, I believe a better approach is to have a path object, instead of a c-string, as the argument of the ctor of the file body, which is how Boost.Iostreams works, e.g. the ctor of `boost::iostreams::mapped_file`. If a `boost::filesystem::path` is the argument, calling `boost::nowide::nowide_filesystem()` in advance is enough to deal with the CJK problem.

**Comment by emmenlau** [comment]
Hey @pdimov , are there further updates on this? Would be great to have full umlaut-support in beast. Thanks a lot for the kind consideration!

**Comment by vinniefalco** [comment]
I thought this was already fixed?

**Comment by pdimov** [comment]
My two PRs seem to have been applied, so the issue should be fixed. Does it still not work?

**Comment by emmenlau** [comment]
Sorry, I missed that this is a full fix! I'll test boost 1.79.0 right now and let you know how it goes!

**Comment by emmenlau** [comment]
Hooray, issue is fixed. Thanks, awesome!
